src/coreclr/System.Private.CoreLib/src/System/Runtime/DependentHandle.cs

```diff
@@ -0,0 +1,242 @@
+                }
+
+                return nGetSecondary(handle);
+            }
+            set
+            {
+                IntPtr handle = _handle;
+
+                if (handle == IntPtr.Zero)
+                {
+                    ThrowHelper.ThrowInvalidOperationException();
+                }
+
+                nSetSecondary(handle, value);
+            }
+        }
+
+        /// <summary>
+        /// Retrieves the values of both <see cref="Target"/> and <see cref="Dependent"/>, if available.
+        /// </summary>
+        /// <returns>The values of <see cref="Target"/> and <see cref="Dependent"/>.</returns>
+        /// <exception cref="InvalidOperationException">Thrown if <see cref="IsAllocated"/> is <see langword="false"/>.</exception>
+        public (object? Target, object? Dependent) GetTargetAndDependent()
+        {
+            IntPtr handle = _handle;
+
+            if (handle == IntPtr.Zero)
+            {
+                ThrowHelper.ThrowInvalidOperationException();
+            }
+
+            object? target = nGetPrimary(handle);
+            object? secondary = nGetSecondary(handle);
+
+            return (target, secondary);
+        }
+
+        /// <summary>
+        /// Gets the target object instance for the current handle.
+        /// </summary>
+        /// <returns>The target object instance, if present.</returns>
+        /// <remarks>This method mirrors <see cref="Target"/>, but without the allocation check.</remarks>
+        internal object? UnsafeGetTarget()
+        {
+            return nGetPrimary(_handle);
+        }
+
+        /// <summary>
+        /// Sets the target object instance for the current handle.
+        /// </summary>
+        /// <remarks>This method mirrors <see cref="Target"/>, but without the allocation check.</remarks>
+        internal void UnsafeSetTarget(object? target)
+        {
+            nSetPrimary(_handle, target);
+        }
+
+        /// <summary>
+        /// Sets the dependent object instance for the current handle.
+        /// </summary>
+        /// <remarks>This method mirrors <see cref="Dependent"/>, but without the allocation check.</remarks>
+        internal void UnsafeSetDependent(object? dependent)
+        {
+            nSetSecondary(_handle, dependent);
+        }
+
+        /// <summary>
+        /// Retrieves the values of both <see cref="Target"/> and <see cref="Dependent"/>, if available.
+        /// </summary>
+        /// <param name="dependent">The dependent instance, if available.</param>
+        /// <returns>The values of <see cref="Target"/> and <see cref="Dependent"/>.</returns>
+        /// <remarks>
+        /// This method mirrors <see cref="GetTargetAndDependent"/>, but without the allocation check.
+        /// The signature is also kept the same as the one for the internal call, to improve the codegen.
+        /// </remarks>
+        internal object? UnsafeGetTargetAndDependent(out object? dependent)
```

Comment:
Do we still need this method? I think the ConditionalWeakTable can call UnsafeGetTarget and UnsafeGetDependent just fine.

Reply:
I think for CoreCLR it would probably be the same, but having `UnsafeGetTargetAndDependent` allows Mono to be a bit more efficient, since we have to perform an array indexing operation there and a couple conditional branches in all cases, and we can cut that in half by using a shared method that returns both:

https://github.com/dotnet/runtime/blob/7a97865c4f2ee9b05aea8605d2b157e951000ac3/src/mono/System.Private.CoreLib/src/System/Runtime/DependentHandle.cs#L117-L144

I think keeping `UnsafeGetTargetAndDependent` might let it be a bit more efficient while keeping the code relatively simpler both in `DependentHandle` and in `ConditionalWeakTable<,>`, so it might be worth keeping it? 🤔

Reply:
Mono implementation of this is less efficient than it needs to be for number of different reasons (e.g. why is the ephemeron array and not just a tuple of values?). I would not worry about it too much.

Reply:
> I think keeping UnsafeGetTargetAndDependent might let it be a bit more efficient

I doubt it. E.g. the value returned by `InternalGetDependent` is actually throw away in some paths. We can avoid calling `InternalGetDependent` completely in some cases.

Reply:
Oh, I see. I've removed `UnsafeGetTargetAndDependent` in 64bad6418ff202c60862eabf22468b508dd662f3 accordingly and made the necessary changes to `ConditionalWeakTable<,>` as well, in theory there shouldn't be regressions at all assuming that jump threading can properly elide those handle `null` checks in there as well. And as you mentioned, in some cases we can now skip getting the dependent entirely, so that's great 😄